### PR TITLE
fix(shell-proxy): support ssh-proxy using http proxy in cygwin

### PR DESCRIPTION
**Pull request by tangruize**
The ssh-proxy.py generates `nc` commands to connect the proxy. However, in cygwin, it does not set the protocol type, leading to `nc` using wrong default protocol (socks5) if we actually use an http proxy.

## Standards checklist:

- [x] The PR title is descriptive.
- [x] The PR doesn't replicate another PR which is already open.
- [x] I have read the contribution guide and followed all the instructions.
- [x] The code follows the code style guide detailed in the wiki.
- [x] The code is mine or it's from somewhere with an MIT-compatible license.
- [x] The code is efficient, to the best of my ability, and does not waste computer resources.
- [x] The code is stable and I have tested it myself, to the best of my abilities.
- [x] If the code introduces new aliases, I provide a valid use case for all plugin users down below.

## Changes:

- add a check ("cygwin") for sys.platform

### CODE REVIEW

**Review by septs — APPROVED** [review]
very good, plz provide tested gif (video)?

**Comment by tangruize** [comment]
Sorry, I don't know how to record a gif. I screenshotted the process (I added a print for the nc commands below):

The proxy setting:

image

After apply the patch this pr ( `if sys.platform in {'linux', 'cygwin'}:`):

image

Before apply the patch (`if sys.platform == 'linux':`):
image

File: plugins/shell-proxy/ssh-proxy.py
Comment (septs) [inline plugins/shell-proxy/ssh-proxy.py:26]: plz update the caveats

**Review by carlosala — COMMENTED** [review]
As soon as we get an approval by @septs I'm happy to merge!

File: plugins/shell-proxy/ssh-proxy.py
Comment (tangruize) [reply plugins/shell-proxy/ssh-proxy.py:26]: I will update the comments tomorrow (because I found that there exists several netcat versions on Linux, such as `netcat-openbsd` (default, works), `netcat-traditional` and `ncat` (nmap netcat)).

Comment (tangruize) [reply plugins/shell-proxy/ssh-proxy.py:26]: I've examined the preinstalled versions of netcat on some Linux distributions, MacOS, and Cygwin. It seems that all of them now support the '-X' option. For a more general check, I think the following modification is better. I have tested it.

```diff
diff --git a/plugins/shell-proxy/ssh-proxy.py b/plugins/shell-proxy/ssh-proxy.py
index 5d239177..e68d109e 100755
--- a/plugins/shell-proxy/ssh-proxy.py
+++ b/plugins/shell-proxy/ssh-proxy.py
@@ -22,7 +22,7 @@ if parsed.scheme not in proxy_protocols:
 
 def make_argv():
     yield "nc"
-    if sys.platform in {'linux', 'cygwin'}:
+    if b'-X ' in subprocess.check_output(['nc', '-h'], stderr=subprocess.STDOUT):
         # caveats: macOS built-in netcat command not supported proxy-type
         yield "-X" # --proxy-type
         # Supported protocols are 4 (SOCKS v4), 5 (SOCKS v5) and connect (HTTP proxy).
```

Comment (tangruize) [reply plugins/shell-proxy/ssh-proxy.py:27]: I've added a caveat.

Comment (tangruize) [reply plugins/shell-proxy/ssh-proxy.py:27]: I also investigated the netcat versions on linux, cygwin and macos, and found that they all support the '-X proto' option. I tried to make it more general by using `if b'-X ' in subprocess.run(['nc', '-h'], check=False, stderr=subprocess.STDOUT, stdout=subprocess.PIPE).stdout:`. However, I failed to clone my repo on MacOS.

**Review by septs — APPROVED** [review]

**Review by carlosala — APPROVED** [review]

**Comment by carlosala** [comment]
Thanks!

**Pull request merged**
